lib/JSON/Validator/Schema/OpenAPIv3.pm

```diff
@@ -0,0 +1,128 @@
+package JSON::Validator::Schema::OpenAPIv3;
+use Mojo::Base 'JSON::Validator::Schema::OpenAPIv2';
+
+use JSON::Validator::Formats;
+use Mojo::Path;
+
+has errors => sub {
+  my $self = shift;
+  my $clone
+    = $self->new(%$self, allow_invalid_ref => 0)->data($self->specification);
+  my @errors = $clone->validate($self->data);
+  return \@errors;
+};
+
+has specification => 'https://spec.openapis.org/oas/3.0/schema/2019-04-02';
+
+sub base_url {
+  my $self = shift;
+  my $data = $self->data;
+
+  # Set
+  if (@_) {
+    my $url    = Mojo::URL->new(shift);
+    my $server = Mojo::URL->new($data->{servers}[0] || '');
+    $server->scheme($url->scheme) if $url->scheme;
+    $server->host($url->host)     if $url->host;
+    $server->port($url->port)     if $url->port;
+    $server->path($url->path)     if $url->path;
+    $data->{servers}[0] = $server->to_string;
+    return $self;
+  }
+
+  # Get
+  my $servers = $data->{servers} || [];
+  return Mojo::URL->new($servers->[0] || '');
+}
+
+sub _build_formats {
+  my $self    = shift;
+  my $formats = $self->SUPER::_build_formats;
+
+  $formats->{'date-time'} = JSON::Validator::Formats->can('check_date_time');
+
+  return $formats;
+}
+
+sub _definitions_path_for_ref {
+  my ($self, $ref) = @_;
+
+  # Try to determine the path from the fqn
+  # We are only interested in the path in the fqn, so following fqn:
+  # "#/components/schemas/some_schema" => ['components', 'schemas']
+  my $path
+    = Mojo::Path->new($ref->fqn =~ m!^.*#/(components/.+)$!)->to_dir->parts;
+  return $path->[0] ? $path : ['components'];
+}
+
+sub _response_schema {
+  my ($self, $schema) = @_;
+  +{
+    description => 'Default Mojolicious::Plugin::OpenAPI response.',
+    content     => {'application/json' => {schema => $schema}},
+  };
+}
+
+sub _sub_schemas         { shift->data->{components}{schemas} ||= {} }
+sub _sub_schemas_pointer {'#/components/schemas'}
+
+sub _validate_type_object {
+  my ($self, $data, $path, $schema) = @_;
+  return shift->SUPER::_validate_type_object(@_) unless ref $data eq 'HASH';
+
+  # "nullable" is the same as "type":["null", ...], which is supported by many
+  # tools, even though not officially supported by OpenAPI.
+  my %properties = %{$schema->{properties} || {}};
+  local $schema->{properties} = \%properties;
+  for my $key (keys %properties) {
+    next unless $properties{$key}{nullable};
+    $properties{$key} = {%{$properties{$key}}};
+    $properties{$key}{type} = ['null', _to_list($properties{$key}{type})];
+  }
+
+  return $self->SUPER::_validate_type_object($_[1], $path, $schema);
+}
+
```

Comment:
Last but not least, missing validation for the requestBody, took the liberty to rewrite from previous code 
 in M::P::OpenAPI::_validate_request_body()

```perl
sub validate_request {
  my ($self, $c, $schema) = @_;
  my @errors = $self->SUPER::validate_request(@_);

  my $ct = $c->req->headers->content_type // '';
  $ct =~ s!;.*$!!;

  my $body_schema = $schema->{requestBody};
  if (my $content = $body_schema->{content}{$ct}) {
    my $body
      = $c->openapi->get_request_data($ct =~ /\bform\b/ ? 'formData' : 'body');
    local $content->{required} //= $body_schema->{required};

    if (my @e = $self->_validate_request_value($content, body => $body)) {
      push @errors, @e;
    }
    else {
      $c->openapi->set_request_data($content, body => $body);
    }
  }
  else {
    push @errors, E '/',
      $ct
      ? "No requestBody rules defined for Content-Type $ct."
      : 'Invalid Content-Type.';
  }
  return @errors;
}
```

Reply:
I've been going back and forth on the validation API today, and I can't figure out a way to make it generic and simple to test. I though adding the helpers to `$c` would make it more generic, but we still have to fiddle around with `$c->req->headers` and other things from `$c`. I don't like it one bit. This is a blocker for this branch: Need a more generic way to validate data. I'm considering splitting it into three steps:

1. J::V::Schema::OpenAPIv3 parses the spec and builds a list of which parameters it can validate. This is a generic data structure such as: `[{in => 'query', name => 'foo'}]` (This step can be cached by J::V::Schema::OpenAPIv3)
2. The list is passed on to M::P::OpenAPI, and M::P::OpenAPI builds builds a new list of parameters that J::V::Schema::OpenAPIv3 can validate. This is a generic data structure such as: `[{in => 'query', name => 'foo', exists => 1, value => 'foo'}]`
3. J::V::Schema::OpenAPIv3 validates the list of data and returns a list of `@errors` that M::P::OpenAPI can serve to the client.

That doesn't sound too bad, but I'm not quite sure how to do the same when it comes to requestBody. The issue here is to make the structure work for both v2 and v3, but maybe it can just be a parameter added to `#1` above? `[{in => 'body', content_type => 'application/json'}, {in => 'body', content_type => 'application/x-www-form-urlencoded'}]` and then the response in `#2` can be: `[{in => 'body', content_type => 'application/json', exists => 1, value => '{"foo":42}'}]`

The code could then be something like, and `$self->validate_request_parameters` should be fairly simple to test.

```
sub validate_request {
  my ($self, $c, $path, $method) = @_;

  # Build a normalized list of all the input data that OpenAPI cares about
  my $parameters = $self->_build_request_parameters($path, $method);

  # Get all the request data from the controller
  my $request = $c->openapi->get_request_data($parameters);

  # _validate_request_parameters() might alter the $request,
  # in case there are default values or some sort of coercion.
  my @errors = $self->_validate_request_parameters($request);

  # Allow the controller to write back request data to $c->param,
  # $c->validation, or whatever.
  $c->openapi->set_request_data($request) unless @errors;

  return @errors;
}
```

But is it flexible enough?

Reply:
I've started implementing it now. Not completely sold yet, but it's better than what I had. I've changed `validate_request()` in OpenAPIv2.pm to the suggested API, added the internal methods to OpenAPIv3.pm, and create a unit test openapiv3-validate.t.